ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/tasks/GetPivotBlockHeaderFromPeerTask.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync.tasks;
+
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.eth.manager.AbstractRetryingPeerTask;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthPeer;
+import tech.pegasys.pantheon.ethereum.eth.manager.exceptions.NoAvailablePeersException;
+import tech.pegasys.pantheon.ethereum.eth.manager.exceptions.PeerBreachedProtocolException;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.LabelledMetric;
+import tech.pegasys.pantheon.metrics.OperationTimer;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class GetPivotBlockHeaderFromPeerTask extends AbstractRetryingPeerTask<List<BlockHeader>> {
+  private static final Logger LOG = LogManager.getLogger();
+  private final ProtocolSchedule<?> protocolSchedule;
+  private final EthContext ethContext;
+  private final LabelledMetric<OperationTimer> ethTasksTimer;
+  private final Optional<EthPeer> peer;
+  private final long pivotBlockNumber;
+
+  private GetPivotBlockHeaderFromPeerTask(
+      final ProtocolSchedule<?> protocolSchedule,
+      final EthContext ethContext,
+      final LabelledMetric<OperationTimer> ethTasksTimer,
+      final Optional<EthPeer> peer,
+      final long pivotBlockNumber,
+      final int maxRetries) {
+    super(ethContext, maxRetries, ethTasksTimer);
+    this.protocolSchedule = protocolSchedule;
+    this.ethContext = ethContext;
+    this.ethTasksTimer = ethTasksTimer;
+    this.peer = peer;
+    this.pivotBlockNumber = pivotBlockNumber;
+  }
+
+  public static GetPivotBlockHeaderFromPeerTask forPivotBlock(
```

Comment:
I would name this task more generically - something like `RetryingGetHeadersFromPeerByNumberTask` since the functionality isn't actually tied to the pivot.  Better yet, you could probably make a completely generic retrying wrapper that takes a supplier for the retryable task.

Reply:
I think a generic retrying wrapper would be nice but I've had a couple of goes and keep hitting issues making it work. It's starting to introduce delegation vs the current inheritance approach which I generally like but it keeps leading to more knock on effects.  I'm not keen to take on that big a change as part of this work and I suspect the better approach is to just move to something like RxJava which appears to give us a lot of benefit.

For now I have renamed this to `RetryingGetHeaderFromPeerByNumberTask`.